flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/slotmanager/DeclarativeSlotManager.java

```diff
@@ -600,61 +603,69 @@ private void allocateSlot(
         FutureUtils.assertNoException(slotAllocationResponseProcessingFuture);
     }
 
-    private void tryFulfillRequirementsWithPendingSlots(
+    private ResourceCounter tryFulfillRequirementsWithPendingSlots(
             JobID jobId,
-            Map<ResourceProfile, Integer> missingResources,
+            Collection<Map.Entry<ResourceProfile, Integer>> missingResources,
             ResourceCounter pendingSlots) {
-        for (Map.Entry<ResourceProfile, Integer> missingResource : missingResources.entrySet()) {
+        for (Map.Entry<ResourceProfile, Integer> missingResource : missingResources) {
             ResourceProfile profile = missingResource.getKey();
             for (int i = 0; i < missingResource.getValue(); i++) {
-                if (!tryFulfillWithPendingSlots(profile, pendingSlots)) {
-                    boolean couldAllocateWorkerAndReserveSlot =
+                final Tuple2<Boolean, ResourceCounter> matchingResult =
+                        tryFulfillWithPendingSlots(profile, pendingSlots);
+                pendingSlots = matchingResult.f1;
+                if (!matchingResult.f0) {
+                    final Tuple2<Boolean, ResourceCounter> allocationResult =
                             tryAllocateWorkerAndReserveSlot(profile, pendingSlots);
+                    pendingSlots = allocationResult.f1;
+                    boolean couldAllocateWorkerAndReserveSlot = allocationResult.f0;
                     if (!couldAllocateWorkerAndReserveSlot && sendNotEnoughResourceNotifications) {
                         LOG.warn("Could not fulfill resource requirements of job {}.", jobId);
                         resourceActions.notifyNotEnoughResourcesAvailable(
                                 jobId, resourceTracker.getAcquiredResources(jobId));
-                        return;
+                        return pendingSlots;
                     }
                 }
             }
         }
+        return pendingSlots;
     }
 
-    private boolean tryFulfillWithPendingSlots(
+    private Tuple2<Boolean, ResourceCounter> tryFulfillWithPendingSlots(
```

Comment:
I would recommend using an explicit type. Something like `MatchingResult` which tells you whether the matching was successful or not. In the former case, it also gives you the newly available pendingSlots back.

The problem with `Tuple2` is that it is not very expressive. You only have `Tuple2.f0` and `Tuple2.f1`. This is not very helpful when someone new to the code wants to understand the code. Having something like `MatchingResult` with `MatchingResult.isSuccess` or `MatchingResult.isSuccessfulMatching` tells you more about the what the intention is.

Reply:
That's a good point.